helix-term/src/ui/picker.rs

```diff
@@ -16,63 +15,72 @@ use tui::widgets::Widget;
+    _preview_view_id: ViewId,
 }
 
-impl<T> PreviewedPicker<T> {
+impl<T> FilePicker<T> {
     pub fn new(
         options: Vec<T>,
         format_fn: impl Fn(&T) -> Cow<str> + 'static,
         callback_fn: impl Fn(&mut Editor, &T, Action) + 'static,
-        preview_fn: impl Fn(&Editor, &T) -> Option<(PathBuf, usize)> + 'static,
+        preview_fn: impl Fn(&Editor, &T) -> Option<FileLocation> + 'static,
     ) -> Self {
         Self {
             picker: Picker::new(options, format_fn, callback_fn),
             preview_cache: HashMap::new(),
-            preview_fn: Box::new(preview_fn),
+            file_fn: Box::new(preview_fn),
+            _preview_view_id: ViewId::default(),
         }
     }
 
-    fn calculate_preview(&mut self, editor: &Editor) {
-        if let Some((path, line)) = self
-            .picker
+    fn current_file(&self, editor: &Editor) -> Option<FileLocation> {
+        self.picker
             .selection()
-            .and_then(|current| (self.preview_fn)(editor, current))
+            .and_then(|current| (self.file_fn)(editor, current))
             .and_then(|(path, line)| canonicalize_path(&path).ok().zip(Some(line)))
-        {
-            let &mut (ref mut doc, ref mut range_map) =
-                self.preview_cache.entry(path.clone()).or_insert_with(|| {
-                    let doc =
-                        Document::open(path, None, Some(&editor.theme), Some(&editor.syn_loader))
-                            .unwrap();
-                    let view = View::new(doc.id());
-                    (doc, hashmap!(line => view))
-                });
-            let view = range_map.entry(line).or_insert_with(|| View::new(doc.id()));
-
-            let range = Range::point(doc.text().line_to_char(line));
-            doc.set_selection(view.id, Selection::from(range));
-            // FIXME: gets aligned top instead of center
-            commands::align_view(doc, view, Align::Center);
+    }
+
+    fn calculate_preview(&mut self, editor: &Editor) {
+        if let Some((path, _line)) = self.current_file(editor) {
+            if !self.preview_cache.contains_key(&path) && editor.document_by_path(&path).is_none() {
+                let mut doc =
+                    Document::open(&path, None, Some(&editor.theme), Some(&editor.syn_loader))
+                        .unwrap();
+                // HACK: a doc needs atleast one selection
+                doc.set_selection(self._preview_view_id, Selection::point(0));
+                self.preview_cache.insert(path, doc);
+            }
         }
     }
 }
 
-impl<T: 'static> Component for PreviewedPicker<T> {
+impl<T: 'static> Component for FilePicker<T> {
     fn render(&self, area: Rect, surface: &mut Surface, cx: &mut Context) {
+        // |---------|  |---------|
+        // |prompt   |  |preview  |
+        // |---------|  |         |
+        // |picker   |  |         |
+        // |         |  |         |
+        // |---------|  |---------|
```

Comment:
Do we have the spaces in between? `|  |`

Reply:
Yeah this is how it looks now:

![Screenshot_2021-08-03_15-23-49](https://user-images.githubusercontent.com/23398472/127996766-3d7836bf-1b69-41ad-be96-50a535b20062.png)

The file picker simply wraps a normal picker and renders it separately before rendering the preview.

Reply:
```suggestion
        // +---------++---------+
        // |prompt   ||preview  |
        // +---------+|         |
        // |picker   ||         |
        // |         ||         |
        // +---------++---------+
```
I think this will look more like it.

Reply:
It does look better, updated :)